webhooks/cluster.go

```diff
@@ -62,22 +68,42 @@ func (c *Cluster) Default(_ context.Context, obj runtime.Object) error {
 }
 
 // ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
-func (c *Cluster) ValidateCreate(_ context.Context, obj runtime.Object) error {
+func (c *Cluster) ValidateCreate(ctx context.Context, obj runtime.Object) error {
+	var allErrs field.ErrorList
+
 	cluster, ok := obj.(*clusterv1.Cluster)
 	if !ok {
 		return apierrors.NewBadRequest(fmt.Sprintf("expected a Cluster but got a %T", obj))
 	}
 
+	// In the case the webhook fails the existing checks return early without waiting for ClusterTopology specific checks.
+
 	// Note: The code in ValidateCreate is intentionally not duplicated to avoid that we accidentally
 	// implement new checks in the API package and forget to duplicate them to the webhook package.
 	// The idea is to add new validation which requires a reader and Cluster variable/patch validation
 	// in the webhook package.
 	// When we drop the method in the API package we must inline it here.
-	return cluster.ValidateCreate()
+	if err := cluster.ValidateCreate(); err != nil {
+		return err
+	}
+
+	// Additional dynamic checks are done if ClusterTopology is enabled and the Cluster has an associated Topology.
+	if feature.Gates.Enabled(feature.ClusterTopology) && cluster.Spec.Topology != nil {
+		if err := c.checkClusterClassExists(ctx, cluster.Namespace, cluster.Spec.Topology.Class); err != nil {
```

Comment:
Q: assuming we have other checks in the pipeline using the ClusterClass (e.g. variable checks), should this be refactored in a way that we can plug in additional checks without re-reading the ClusterClass?
(happy to talk about this if this is not clear)

Reply:
Yeah - it's just a question of whether to do this in this PR or a follow up - it won't be a major change either way, but we might want to think about how we structure the API calls in this webhook - there's going to be a number of them.